packages/flutter/lib/src/material/popup_menu.dart

```diff
@@ -460,37 +460,76 @@ class _PopupMenu<T> extends StatelessWidget {
   }
 }
 
+// Positioning of the menu on the screen.
 class _PopupMenuRouteLayout extends SingleChildLayoutDelegate {
-  _PopupMenuRouteLayout(this.position, this.selectedItemOffset);
+  _PopupMenuRouteLayout(this.position, this.selectedItemOffset, this.textDirection);
 
+  // Rectangle of underlying button, relative to the overlay's dimensions.
   final RelativeRect position;
+
+  // The distance from the top of the menu to the middle of selected item.
+  //
+  // This will be null if there's no item to position in this way.
   final double selectedItemOffset;
 
+  // Whether to prefer going to the left or to the right.
+  final TextDirection textDirection;
+
+  // We put the child wherever position specifies, so long as it will fit within
+  // the specified parent size padded (inset) by 8. If necessary, we adjust the
+  // child's position so that it fits.
+
   @override
   BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
-    return constraints.loosen();
+    // The menu can be at most the size of the overlay minus 8.0 pixels in each
+    // direction.
+    return new BoxConstraints.loose(constraints.biggest - const Offset(_kMenuScreenPadding * 2.0, _kMenuScreenPadding * 2.0));
   }
 
-  // Put the child wherever position specifies, so long as it will fit within the
-  // specified parent size padded (inset) by 8. If necessary, adjust the child's
-  // position so that it fits.
   @override
   Offset getPositionForChild(Size size, Size childSize) {
-    double x = position?.left
-      ?? (position?.right != null ? size.width - (position.right + childSize.width) : _kMenuScreenPadding);
-    double y = position?.top
-      ?? (position?.bottom != null ? size.height - (position.bottom - childSize.height) : _kMenuScreenPadding);
+    // size: The size of the overlay.
+    // childSize: The size of the menu, when fully open, as determined by
+    // getConstraintsForChild.
+
+    // Find the ideal vertical position.
+    double y;
+    if (selectedItemOffset == null) {
+      y = position.top;
+    } else {
+      y = position.top + (size.height - position.top - position.bottom) / 2.0 - selectedItemOffset;
```

Comment:
Could you explain why you need to add `(size.height - position.top - position.bottom) / 2.0` to the top offset? It makes my popup menu showing at a wrong place...

Reply:
If you're seeing a problem with popup menus, please open a new issue and include a small test case that demos the problem. Happy to take a look!